sale_coupon_multi_use/models/sale_order.py

```diff
@@ -1,32 +1,134 @@
-from odoo import models
+from odoo import fields, models
+
+ORDER_CTX_KEY = "coupon_sale_order"
+
+
+def _get_order_line_coupons_filter(line):
+    return lambda r: r.program_id.discount_line_product_id == line.product_id
 
 
 class SaleOrder(models.Model):
     """Extend to modify action_confirm for multi-use coupons."""
 
     _inherit = "sale.order"
 
-    # TODO: implement multi-use coupon relation on sale.order ->
-    # applied_coupon_ids is o2m, which removes coupon if same coupon
-    # is applied on another SO. But for multi-use coupon, it must be
-    # able to keep same coupon on multiple sale orders.
-    # TODO: action_draft seems buggy, because it does not reset back
-    # coupons to new state (you can abuse coupon this way and use normal
-    # coupon multiple times).
+    coupon_multi_use_ids = fields.Many2many(
+        "sale.coupon",
+        "sale_order_coupon_multi_rel",
+        "sale_id",
+        "coupon_id",
+        string="Multi Use Coupons",
+        copy=False,
+        readonly=True,
+    )
+
+    def _get_multi_use_coupons_to_handle(self, filter_func=None):
+        self.ensure_one()
+        coupons = self.coupon_multi_use_ids - self.applied_coupon_ids
+        if filter_func:
+            return filter_func(coupons)
+        return coupons
+
+    def _trigger_multi_use_coupons(self, state, filter_func=None):
+        self.ensure_one()
+        # NOTE. This method must be called with ORDER_CTX_KEY on order.
+        multi_use_coupons = self._get_multi_use_coupons_to_handle()
+        if multi_use_coupons:
+            multi_use_coupons.write({"state": state})
+        return multi_use_coupons
 
     def action_confirm(self):
         """Extend to pass coupon_sale_order context."""
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            # Mimic same behavior as for single-user coupon.
+            order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_confirm()
+            order._trigger_multi_use_coupons("used")
```

Comment:
Use a function on sale coupon to delegate the state management, by example `use_coupon()`